### ISSUE TITLE
Create a custom function using the functions from math_cpu.cc source code in Caffe2?

### ISSUE THREAD

**Issue by d-b-h** [pytorch/pytorch#12045]
I would like to write my own convolution function in C++ using the GEMM functions defined in math_cpu.cc.  I am planning to write a python wrapper for the function and use in a python environment.  Is this possible?  The reason for this is that the dependencies are already present with the installation of caffe2.

**Comment by steplee** [comment]
Yes, you can put it in the source tree in `caffe2/operators`, or build a shared object/archive to link against.

It gets wrapped in python automatically. If you go the shared object route and want python access, you need to import it and, I think, call `caffe2.python.core.RefreshRegisteredOperators()`.

**Comment by d-b-h** [comment]
How is creating an operator for convolution different than writing a function that does the same?  I am unaware of the required linking and building of dependencies involved in doing such a task.

**Comment by steplee** [comment]
The purpose of a deep learning framework is to wire together a computational graph that does forward and backward passes. To compute gradients for the backward pass, you need more than a typical C++ function that only returns the answer during a forward pass.

The difference between the two is that an operator wraps a function, provides another operator for its gradient, and, crucially, can be used as a node in a net. 

See [this link](https://caffe2.ai/docs/custom-operators.html). Also, see the hundreds of ops from `pytorch/caffe2/operators`. If you want to build the op out of the source tree, I have two repos in my profile that may help you get started (but they are messy and I haven't written gradient ops yet).

**Comment by d-b-h** [comment]
I understand, thank you.  I looked at a couple of operators on your profile and looked at the references you have linked to as well. Can you recommend a starting point for a complete beginner in caffe2 trying to write their own operator?
